File: Leetcode - Top Interview Questions/trappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) 
    {
        int n=height.size();
      //  vector<int> ans;
        vector<int> left(n,0);
        vector<int> right(n,0);
        left[0]=height[0];
        right[n-1]=height[n-1];
        //traverse left and right 
        if(n<=2)
        {
            return 0;
        }
        for(int i=1;i<n;i++)
        {
            left[i]=max(left[i-1],height[i]);           
        }
        for(int i=1;i<n;i++)
        {
            right[n-i-1]=max(right[n-i],height[n-i-1]); 
        }
        int rain=0;
        for(int i=0;i<n;i++)
        {
            rain += min(left[i],right[i])-height[i];
         //   ans.push_back(rain);
        }
       // return ans.size()-1;
        return rain;
    }
};

File: Leetcode - Top Interview Questions/TrappingRainWater.java
/**
 * TrappingRainWater
 * https://leetcode.com/problems/trapping-rain-water/
 */
public class TrappingRainWater {
    int trap(int[] height){
        int n = height.length, ans = 0, maxi = 0, mini = 0;
        // two pointers 
        int i = 0, j = n - 1;
        // pointer i from the left
        // pointer j from the right
        while(i <= j){
            mini = Math.min(height[i], height[j]);
            maxi = Math.max(maxi, mini);
            
            // the units of water being tapped is the diffence between max height and min height
            ans += maxi - mini;
            if(height[i] < height[j])i++;
            else j--;
        }
        return ans;
    }
}
    